<?xml version="1.0"?>
<launch>
  <arg name="camera_name"   default="/rgb"/>
  <arg name="allow_growth"  default="false"/>
  <arg name="show_rviz"     default="false"/>

  <!-- people detection -->
  <arg name="cpu"                   default="false"/>
  <arg name="pose_2d_detect"        default="true"/>
  <arg name="pose_2d_img_show"      default="false"/>
  <arg name="pose_2d_img_pub"       default="true"/>
  <arg name="pose_2d_log_show"      default="false"/>

  <!-- detection parameters -->
  <arg name="detection_confidence_thresh"           default="0.1"/>
  <arg name="detection_border_thresh_w"             default="100"/>
  <arg name="detection_border_thresh_h"             default="25"/>

  <!-- UKF parameters -->
  <arg name="measurement_noise_pix_cov"             default="100"/>
  <arg name="process_noise_pos_cov"                 default="0.03"/>
  <arg name="process_noise_vel_cov"                 default="0.01"/>
  <arg name="process_noise_height_cov"              default="1e-10"/>

  <!-- tracking parameters -->
  <arg name="init_cov_scale"                        default="0.25"/>
  <arg name="association_maha_sq_thresh"            default="9.0"/>
  <arg name="association_neck_ankle_max_dist"       default="200"/>
  <arg name="association_neck_max_dist"             default="150"/>
  <arg name="tracking_remove_trace_thresh"          default="2.0"/>
  <arg name="tracking_newtrack_dist2exists_thersh"  default="100"/>

  <!-- person identification -->
  <arg name="imprinting_max_dist"                   default="4.0"/>
  <arg name="initial_training_num_samples"          default="10"/>
  <arg name="min_target_confidence"                 default="0.1"/>
  <arg name="id_switch_detection_thresh"            default="-0.1"/>
  <arg name="reid_confidence_thresh"                default="0.1"/>
  <arg name="reid_positive_count"                   default="5"/>


  <!-- *** people detection *** -->
  <include file="$(find lightweight_human_pose_estimation)/launch/human_pose_2d.launch">
    <param name="cpu"                   type="bool"     value="$(arg cpu)"/>
    <param name="input_image_topic"     type="str"      value="$(arg camera_name)/image_raw"/>

    <param name="pose_2d_detect"        type="bool"     value="$(arg pose_2d_detect)"/>
    <param name="pose_2d_img_show"      type="bool"     value="$(arg pose_2d_img_show)"/>
    <param name="pose_2d_img_pub"       type="bool"     value="$(arg pose_2d_img_pub)"/>
    <param name="pose_2d_log_show"      type="bool"     value="$(arg pose_2d_log_show)"/>
  </include>

  <!-- *** people tracking *** -->
  <node pkg="monocular_people_tracking" type="monocular_people_tracking_node" name="monocular_people_tracking" output="screen">
    <remap from="camera_info" to="$(arg camera_name)/camera_info"/>

    <!-- detection parameters -->
    <param name="detection_confidence_thresh"           value="$(arg detection_confidence_thresh)"/>
    <param name="detection_border_thresh_w"             value="$(arg detection_border_thresh_w)"/>
    <param name="detection_border_thresh_h"             value="$(arg detection_border_thresh_h)"/>

    <!-- UKF parameters -->
    <param name="measurement_noise_pix_cov"             value="$(arg measurement_noise_pix_cov)"/>
    <param name="process_noise_pos_cov"                 value="$(arg process_noise_pos_cov)"/>
    <param name="process_noise_vel_cov"                 value="$(arg process_noise_vel_cov)"/>
    <param name="process_noise_height_cov"              value="$(arg process_noise_height_cov)"/>

    <!-- tracking parameters -->
    <param name="init_cov_scale"                        value="$(arg init_cov_scale)"/>
    <param name="association_maha_sq_thresh"            value="$(arg association_maha_sq_thresh)"/>
    <param name="association_neck_ankle_max_dist"       value="$(arg association_neck_ankle_max_dist)"/>
    <param name="association_neck_max_dist"             value="$(arg association_neck_max_dist)"/>
    <param name="tracking_remove_trace_thresh"          value="$(arg tracking_remove_trace_thresh)"/>
    <param name="tracking_newtrack_dist2exists_thersh"  value="$(arg tracking_newtrack_dist2exists_thersh)"/>
  </node>

  <!-- *** person identification *** -->
  <node pkg="monocular_person_following" type="monocular_person_following_node" name="monocular_person_following" output="screen">
    <remap from="image" to="$(arg camera_name)/image_raw"/>

    <!--
    /// @brief
    /// Initial state:
    ///   if there is a person in front of the camera (within imprinting_max_dist),
    ///   the person is registered as the target
    /// Initial Training state:
    ///   the target person features are added to the classifier a certain time (initial_training_num_samples),
    ///   then, the sytem transits to the tracking state
    /// Tracking state:
    ///   if the identification confidence of the target is lower than min_target_confidence,
    ///   the system judges that the target is lost, and transits to ReID state
    /// ReID state:
    ///   if a track shows a confidence higher than reid_confidence_thresh several times (reid_positive_count),
    ///   the track is reidentified as the target, and the system transits to Tracking state
    -->
    <param name="imprinting_max_dist"           value="$(arg imprinting_max_dist)"/>
    <param name="initial_training_num_samples"  value="$(arg initial_training_num_samples)"/>
    <param name="min_target_confidence"         value="$(arg min_target_confidence)"/>
    <param name="id_switch_detection_thresh"    value="$(arg id_switch_detection_thresh)"/>
    <param name="reid_confidence_thresh"        value="$(arg reid_confidence_thresh)"/>
    <param name="reid_positive_count"           value="$(arg reid_positive_count)"/>
  </node>

  <!-- *** visualization *** -->
  <!-- <node pkg="monocular_person_following" type="visualization.py" name="visualization_node" output="screen"> -->
  <node pkg="lightweight_human_pose_estimation" type="visualization.py" name="visualization_node" output="screen">
    <!-- <remap from="image_rect" to="$(arg camera_name)/image_rect"/> -->
    <remap from="image_raw" to="$(arg camera_name)/image_raw"/>
    <param name="show"      value="$(arg show_rviz)"/>
  </node>
  <node pkg="image_transport" type="republish" name="compress_visualize" args="raw in:=/visualization_node/visualize compressed out:=/visualization_node/visualize"/>
  <node pkg="topic_tools"     type="throttle"  name="throttle_visualize" args="messages /visualization_node/visualize/compressed 4 /visualization_node/visualize_slow/compressed"/>

  <!-- *** gesture recognition *** -->
  <node pkg="monocular_person_following" type="simple_gesture_recognition.py" name="simple_gesture_recognition" output="screen"/>

  <!-- *** robot controller *** -->
  <!-- <node pkg="monocular_person_following" type="robot_controller.py" name="robot_controller" output="screen" if="false">
    <remap from="cmd_vel" to="/RosAria/cmd_vel"/>
    <param name="enable_back" value="false"/>
    <param name="max_vx" value="0.1"/>
    <param name="max_va" value="0.1"/>
    <param name="gain_vx" value="0.0"/>
    <param name="gain_va" value="0.1"/>
    <param name="distance" value="2.5"/>
    <param name="timeout" value="0.5"/>
  </node> -->
</launch>
